audio/out/ao_oss.c

```diff
@@ -0,0 +1,440 @@
+            return; /* Avoid infinite loop: init() -> reset() -> init()... */
+        p->resetting = true;
+        MP_WARN(ao, "Force reinitialize audio device.\n");
+        uninit(ao);
+        init(ao);
+        p->resetting = false;
+    }
+}
+
+static bool set_pause(struct ao *ao, bool pause)
+{
+    struct priv *p = ao->priv;
+    int trig = 0; /* Both directions disabled */
+
+    if (!pause) {
+        trig |= (PCM_ENABLE_INPUT | PCM_ENABLE_OUTPUT); /* Both directions enabled */
+    }
+    if (ioctl(p->dsp_fd, SNDCTL_DSP_SETTRIGGER, &trig) != 0) {
+        p->playing = false;
+        MP_WARN_IOCTL_ERR(ao);
+    }
+
+    return true;
+}
+
+static void start(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+
+    p->playing = true;
+    set_pause(ao, false);
+}
+
+static bool audio_write(struct ao *ao, void **data, int samples)
+{
+    struct priv *p = ao->priv;
+    const size_t size = (samples * ao->sstride);
+    ssize_t rc;
+
+    if (size == 0)
+        return true;
+
+    rc = write(p->dsp_fd, data[0], size);
+    if (rc == -1) {
+        MP_WARN(ao, "audio_write: write() fail, err = %i: %s\n",
+            errno, strerror(errno));
+        return false;
+    }
+    if ((size_t)rc != size) {
+        p->playing = false;
+        MP_WARN(ao, "audio_write: unexpected partial write: required: %zu, writed: %zu\n",
+            size, (size_t)rc);
+        return false;
+    }
+
+    return true;
+}
+
+static void get_state(struct ao *ao, struct mp_pcm_state *state)
+{
+    struct priv *p = ao->priv;
+    audio_buf_info info;
+    int odelay;
+
+    /* free_samples, queued_samples */
+    if (ioctl(p->dsp_fd, SNDCTL_DSP_GETOSPACE, &info) != -1) {
+        state->free_samples = (info.bytes / ao->sstride);
+        state->queued_samples = (ao->device_buffer - state->free_samples);
+    } else {
+        state->free_samples = state->queued_samples = 0;
+        p->playing = false;
+        MP_WARN_IOCTL_ERR(ao);
+    }
+    /* delay */
+    if (ioctl(p->dsp_fd, SNDCTL_DSP_GETODELAY, &odelay) != -1) {
+        state->delay = (odelay / (double)ao->bps);
+    } else {
+        state->delay = 0;
+        p->playing = false;
+        MP_WARN_IOCTL_ERR(ao);
+    }
+    /* playing */
+    state->playing = p->playing;
+}
+
+static void list_devs(struct ao *ao, struct ao_device_list *list)
+{
+    struct stat st;
+    struct ao_device_desc dev;
+    size_t i, buf_size, fail_cnt;
+    char dev_path[32], dev_descr[256];
+
+    if (stat(PATH_DEV_DSP, &st) == 0) {
+        dev.name = PATH_DEV_DSP;
+        dev.desc = "Default";
+        ao_device_list_add(list, ao, &dev);
+    }
+
+    /* Auto detect. */
+    dev.name = dev_path;
+    dev.desc = dev_descr;
+    for (i = 0, fail_cnt = 0; fail_cnt < 8; i ++) {
+        /* Get device description. */
+        snprintf(dev_path, sizeof(dev_path), "dev.pcm.%zu.%%desc", i);
+        buf_size = (sizeof(dev_descr) - 1);
+        if (sysctlbyname(dev_path, dev_descr, &buf_size, NULL, 0) != 0) {
```

Comment:
Do you plan to support non-BSDs e.g., Solaris and 4Front reference? If so sprinkle `#if defined(__DragonFly__) || defined(__FreeBSD__)`, otherwise add `(os-dragonfly || os-freebsd)` to `deps` in wscript.

Reply:
Does DragonFlyBSD have ```dev.pcm.%zu.%desc``` sysctls ?

Reply:
I think so, see https://github.com/dragonflybsd/dragonflybsd/commit/8e4d13764bdfc916c23fb22158c65e486dce631c. Maybe @zrj-rimwis or @tuxillo can confirm.

Reply:
I think so, see an example in my box:

```
$ sysctl dev.pcm
dev.pcm.0.%desc: Intel (0x2808) (HDMI/DP 8ch)
dev.pcm.0.%driver: pcm
dev.pcm.0.%location: nid=3
dev.pcm.0.%parent: hdaa0
dev.pcm.0.play.vchans: 1
dev.pcm.0.play.vchanmode: passthrough
dev.pcm.0.play.vchanrate: 48000
dev.pcm.0.play.vchanformat: s16le:2.0
dev.pcm.0.play.32bit: 24
dev.pcm.0.buffersize: 65536
dev.pcm.0.bitperfect: 0
dev.pcm.1.%desc: Realtek (0x0283) (Right Analog Headphones)
dev.pcm.1.%driver: pcm
dev.pcm.1.%location: nid=33
dev.pcm.1.%parent: hdaa1
dev.pcm.1.play.vchans: 2
dev.pcm.1.play.vchanmode: fixed
dev.pcm.1.play.vchanrate: 48000
dev.pcm.1.play.vchanformat: s16le:2.0
dev.pcm.1.play.32bit: 24
dev.pcm.1.buffersize: 65536
dev.pcm.1.bitperfect: 0
```